executor/testdata/executor_suite_out.json

```diff
@@ -598,5 +598,950 @@
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Res": [
+          "1999998 1999998 Filler ...",
+          "1999999 1999999 Filler ..."
+        ]
+      },
+      {
+        "SQL": "SELECT * FROM t WHERE a IN (1999998, 1999999, 2000000)",
+        "Plan": [
+          "TableReader 30.00 root partition:p1,p2 data:Selection",
+          "└─Selection 30.00 cop[tikv]  in(testrangepartitionboundaries.t.a, 1999998, 1999999, 2000000)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Res": [
+          "1999998 1999998 Filler ...",
+          "1999999 1999999 Filler ...",
+          "2000000 2000000 Filler ..."
+        ]
+      },
+      {
+        "SQL": "SELECT * FROM t WHERE a IN (1999999, 2000000, 2000001)",
+        "Plan": [
+          "TableReader 30.00 root partition:p1,p2 data:Selection",
+          "└─Selection 30.00 cop[tikv]  in(testrangepartitionboundaries.t.a, 1999999, 2000000, 2000001)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Res": [
+          "1999999 1999999 Filler ...",
+          "2000000 2000000 Filler ...",
+          "2000001 2000001 Filler ..."
+        ]
+      },
+      {
+        "SQL": "SELECT * FROM t WHERE a IN (2000000, 2000001, 2000002)",
+        "Plan": [
+          "TableReader 30.00 root partition:p2 data:Selection",
+          "└─Selection 30.00 cop[tikv]  in(testrangepartitionboundaries.t.a, 2000000, 2000001, 2000002)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Res": [
+          "2000000 2000000 Filler ...",
+          "2000001 2000001 Filler ...",
+          "2000002 2000002 Filler ..."
+        ]
+      },
+      {
+        "SQL": "SELECT * FROM t WHERE a IN (2999997, 2999998, 2999999)",
+        "Plan": [
+          "TableReader 30.00 root partition:p2 data:Selection",
+          "└─Selection 30.00 cop[tikv]  in(testrangepartitionboundaries.t.a, 2999997, 2999998, 2999999)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Res": [
+          "2999998 2999998 Filler ...",
+          "2999999 2999999 Filler ..."
+        ]
+      },
+      {
+        "SQL": "SELECT * FROM t WHERE a IN (2999998, 2999999, 3000000)",
+        "Plan": [
+          "TableReader 30.00 root partition:p2 data:Selection",
+          "└─Selection 30.00 cop[tikv]  in(testrangepartitionboundaries.t.a, 2999998, 2999999, 3000000)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Res": [
+          "2999998 2999998 Filler ...",
+          "2999999 2999999 Filler ..."
+        ]
+      },
+      {
+        "SQL": "SELECT * FROM t WHERE a IN (2999999, 3000000, 3000001)",
```

Comment:
How about adding some cases mixing them with some values in other types?
`a IN (2999999, 3000000, 3000001, 2999999.9, 'abcde')`

Reply:
I think that would only test the type conversion layer in TiDB, not the partitioning pruning which should happen after the type conversion is done.

A quick test shows:
```sql
explain select * from t where a IN (3000000, 3000001, '2999999', 2999999.9, 'abcde')
--------------

+------------------------------+---------+-----------+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                           | estRows | task      | access object         | operator info                                                                                                                                             |
+------------------------------+---------+-----------+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------+
| PartitionUnion_10            | 11.21   | root      |                       |                                                                                                                                                           |
| ├─TableReader_13             | 3.20    | root      |                       | data:Selection_12                                                                                                                                         |
| │ └─Selection_12             | 3.20    | cop[tikv] |                       | or(or(eq(test.t.a, 3000000), eq(test.t.a, 3000001)), or(eq(test.t.a, 2999999), or(eq(cast(test.t.a, decimal(20,0) BINARY), 2999999.9), eq(test.t.a, 0)))) |
| │   └─TableFullScan_11       | 4.00    | cop[tikv] | table:t, partition:p0 | keep order:false, stats:pseudo                                                                                                                            |
| ├─TableReader_16             | 4.00    | root      |                       | data:Selection_15                                                                                                                                         |
| │ └─Selection_15             | 4.00    | cop[tikv] |                       | or(or(eq(test.t.a, 3000000), eq(test.t.a, 3000001)), or(eq(test.t.a, 2999999), or(eq(cast(test.t.a, decimal(20,0) BINARY), 2999999.9), eq(test.t.a, 0)))) |
| │   └─TableFullScan_14       | 5.00    | cop[tikv] | table:t, partition:p1 | keep order:false, stats:pseudo                                                                                                                            |
| └─TableReader_19             | 4.00    | root      |                       | data:Selection_18                                                                                                                                         |
|   └─Selection_18             | 4.00    | cop[tikv] |                       | or(or(eq(test.t.a, 3000000), eq(test.t.a, 3000001)), or(eq(test.t.a, 2999999), or(eq(cast(test.t.a, decimal(20,0) BINARY), 2999999.9), eq(test.t.a, 0)))) |
|     └─TableFullScan_17       | 5.00    | cop[tikv] | table:t, partition:p2 | keep order:false, stats:pseudo                                                                                                                            |
+------------------------------+---------+-----------+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------+
10 rows in set, 2 warnings (0.00 sec)

tidb> show warnings;
--------------
show warnings
--------------

+---------+------+------------------------------------------+
| Level   | Code | Message                                  |
+---------+------+------------------------------------------+
| Warning | 1292 | Truncated incorrect FLOAT value: 'abcde' |
| Warning | 1292 | Truncated incorrect FLOAT value: 'abcde' |
+---------+------+------------------------------------------+
2 rows in set (0.00 sec)
```
Which does point out some of issues:
- We are not generating correct CAST syntax; should be CAST(test.t.a AS decimal(20,0))
- We are casting the table value, instead of casting the predicate (WHERE value). Not sure which is the correct way?
- The expression to evaluate could be pruned, since some of the OR predicates cannot match any rows in some specific partition. I.e. partition p2 only has `a` values between 2000000 and 2999999 and can never match values outside of that range, so no need to even try to evaluate them.

Should I create new issues for these with label 'type/enhancement' or add such test cases to this PR?

Reply:
For the first and the second issue, the behavior of casting in TiDB may be different with MySQL, and discussing it here may be a little beyond this PR's scope. Most casting actions are handled in this function `newBaseBuiltinFuncWithTp`, and you can read it if you'd like to. I agree with you that these cases only test the conversation layer, so how about just ignore them here?

For the third issue, you can `set tidb_partition_prune_mode='dynamic'` before processing the query, then the problem will not exist.

Reply:
I agree, let us just ignore the casting tests in this PR.

I added `set tidb_partition_prune_mode='dynamic'` to the test.